### ISSUE TITLE
[Ecore Editor] Fix various issues

### ISSUE THREAD

**Issue by MWEclipse** [MWEclipse/testing#187]
| --- | --- |
| Bugzilla Link | [544774](https://bugs.eclipse.org/bugs/show_bug.cgi?id=544774) |
| Status | NEW |
| Importance | P3 normal |
| Reported | Feb 25, 2019 09:02 EDT |
| Modified | Aug 04, 2023 01:28 EDT |
| Depends on | [480729](https://bugs.eclipse.org/bugs/show_bug.cgi?id=480729), [487113](https://bugs.eclipse.org/bugs/show_bug.cgi?id=487113), [495176](https://bugs.eclipse.org/bugs/show_bug.cgi?id=495176), [501256](https://bugs.eclipse.org/bugs/show_bug.cgi?id=501256), [501257](https://bugs.eclipse.org/bugs/show_bug.cgi?id=501257), [501260](https://bugs.eclipse.org/bugs/show_bug.cgi?id=501260) |
| Reporter | Lucas Koehler |

## Description 
The following issues should be fixed in the EMFForms Ecore Editor:

1. Too many selection options in EKeys control\
The EKeys control of an EReference allows to select any EAttribute in the ecore but only attributes of the EReference's type are allowed.\
Also the create and link button creates the attribute in the wrong place: The attribute is created in the EClass containing the EReference. It should be created in the EReference's type or be not allowed at all (Not allowed at all is probably better)

2. EOperation view - issues with EParameters control:
- Parameters cannot be edited after they have been created. 
- Pressing create and link creates an empty parameter. Due to the point above it cannot be created afterwards.
- Link button makes no sense because parameters are contained
- The order of the parameters cannot be changed.\
Basically the same issues apply for the ETypeParameters' control

3. Code Generation - insufficient feedback:\
Not enough feedback is given if nothing could be generated. E.g. I tried to generate tests bundle for a modified coffee model: nothing was generated and no error was reported. The progress dialog just closed after a while.

**Comment by MWEclipse** [comment]
By Eugen Neufeld on Dec 03, 2019 11:00

MAJOR:\
Changing the min or max bounds via the "+" or "-" button breaks the min/max bounds for all attributes / references until editor is re-opened.\
It's not possible to add the type parameters for an operation parameter. In this case we can't add the type parameters for the "context" parameter of the "hasName" operation. The Ecore Editor showns that as a warning permanently

MINOR:\
Linking an EOpposite sets both references, however it shows a false positive validation error which persists until editor is closed and reopened\
Deleting an EOpposite link removes only one of two references, however it doesn't show a validation error until editor is closed and reopened\
When creating new enums, the enum table will show a false positive error validation icon until the editor is closed and re-opened\
The "Default Value Literal" can be set, even in cases where it doesn't make sense, for example the "assignee" reference in "Task"\
In Enum Creation Popup: "Validation" Column is too large.\
In Enum Creation Popup and in Enum View: Name field should be focused when adding a new literal\
Selecting an EType from a popup which is error validated should not paint all possible values as red.\
"New sibling" would be a nice feature, similar to the normal editor\
BaseType Annotation offers enum support to select basetype in normal editor. Would be nice to have in Ecore Editor too, instead of specifying it by hand

INFO:\
Task Package: Why is Ns URI shown before Ns Prefix? Not consistent to normal editor.\
Properties view is empty, except when clicking on links it will show the info of the ecore file. Could be used better?

**Comment by MWEclipse** [comment]
By Paul Reuter on Aug 12, 2021 20:02

*** [Bug 575363](https://bugs.eclipse.org/bugs/show_bug.cgi?id=575363) has been marked as a duplicate of this bug. ***

**Comment by MWEclipse** [comment]
Aug 04, 2023 01:28

This bug hasn't had any activity in quite some time. Maybe the problem got resolved, was a duplicate of something else, or became less pressing for some reason - or maybe it's still relevant but just hasn't been looked at yet.

If you have further information on the current state of the bug, please add it. The information can be, for example, that the problem still occurs, that you still want the feature, that more information is needed, or that the bug is (for whatever reason) no longer relevant.

--\
The automated Eclipse Genie.
